core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1263,17 +1241,34 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
             "trying again...")
           store.close()
           memoryManager.release(appId, attempt.info.attemptId)
-          lease.rollback()
           retried = true
-
         case e: Exception =>
           store.close()
           memoryManager.release(appId, attempt.info.attemptId)
-          lease.rollback()
           throw e
       }
     }
 
+    // Create a LevelDB and start a background thread to dump data to LevelDB
+    logInfo(s"Leasing disk manager space for app $appId / ${attempt.info.attemptId}...")
```

Comment:
There are still chances here exceptions have been thrown, and then we’ll leak HybridStore instance with memory being allocated, and never be released.

Reply:
do you mean exception thrown in the foreground thread or in the background thread? If it's in foreground thread, we can use a try-catch block to close the store when an exception is thrown.

Reply:
Foreground, before switching method is called. Yeah I meant that. Thanks.

Reply:
ok, will update it. Thanks